Python/pystate.c

```diff
@@ -1386,28 +1391,62 @@ allocate_chunk(int size_in_bytes, _PyStackChunk* previous)
     return res;
 }
 
-static _PyThreadStateImpl *
-alloc_threadstate(void)
+static void
+reset_threadstate(_PyThreadStateImpl *tstate)
 {
-    return PyMem_RawCalloc(1, sizeof(_PyThreadStateImpl));
+    // Set to _PyThreadState_INIT directly?
+    memcpy(tstate,
+           &initial._main_interpreter._initial_thread,
+           sizeof(*tstate));
+}
+
+#define LOCK_THREADS_FREELIST(interp) \
+    PyMutex_LockFlags(&(interp)->threads.freelist.mutex, _Py_LOCK_DONT_DETACH)
+#define UNLOCK_THREADS_FREELIST(interp) \
+    PyMutex_Unlock(&(interp)->threads.freelist.mutex)
+
+static _PyThreadStateImpl *
+alloc_threadstate(PyInterpreterState *interp)
+{
+    // Try the freelist first.
+    _PyThreadStateImpl *tstate = NULL;
+    LOCK_THREADS_FREELIST(interp);
+    if (interp->threads.freelist.head != NULL) {
+        tstate = interp->threads.freelist.head;
+        interp->threads.freelist.head =
+            (_PyThreadStateImpl *)tstate->base.next;
+    }
+    UNLOCK_THREADS_FREELIST(interp);
```

Comment:
Regarding trying to do this without a lock, we could do something like:

```c
tstate = _Py_atomic_exchange_ptr(&interp->threads.freelist.head, NULL);
```

I'm not totally sure what the performance benefit would be, but I *think* it should be a little better than a lock.

Reply:
It may be.  If I switch back to a dedicated field for just the one thread state, instead of a freelist, then I'll definitely go with the atomic compare-and-swap.

Reply:
done